### ISSUE TITLE
ENGINE_add() failure

### ISSUE THREAD

**Issue by RoyZhang2022** [openssl/openssl#19010]
I am using AWS C++ SDK to develop my application. My repository is on github. AWS SDK works fine for one month. But from last Wednesday AWS SDK initAPI() started to fail. Based on the error info and the debug info I added I have identified the line that introduces the error. It is from s2n-tls code ENGINE_add(e) of s2n_rand_init(). As ENGINE_add(e) is an openssl function I checked its implementation.

--s2n-tls code
```
370 S2N_RESULT s2n_rand_init(void)
371 {
372     RESULT_GUARD_POSIX(s2n_rand_init_cb());
373
374     RESULT_GUARD(s2n_ensure_initialized_drbgs());
375
376 #if S2N_LIBCRYPTO_SUPPORTS_CUSTOM_RAND
377     /* Create an engine */
378     ENGINE *e = ENGINE_new();
379
380     RESULT_ENSURE(e != NULL, S2N_ERR_OPEN_RANDOM);
381     RESULT_GUARD_OSSL(ENGINE_set_id(e, S2N_RAND_ENGINE_ID), S2N_ERR_OPEN_RANDOM);
382     RESULT_GUARD_OSSL(ENGINE_set_name(e, "s2n entropy generator"), S2N_ERR_OPEN_RANDOM);
383     RESULT_GUARD_OSSL(ENGINE_set_flags(e, ENGINE_FLAGS_NO_REGISTER_ALL), S2N_ERR_OPEN_RANDOM);
384     RESULT_GUARD_OSSL(ENGINE_set_init_function(e, s2n_openssl_compat_init), S2N_ERR_OPEN_RANDOM);
385     RESULT_GUARD_OSSL(ENGINE_set_RAND(e, &s2n_openssl_rand_method), S2N_ERR_OPEN_RANDOM);
**386     RESULT_GUARD_OSSL(ENGINE_add(e), S2N_ERR_OPEN_RANDOM);**
387     RESULT_GUARD_OSSL(ENGINE_free(e) , S2N_ERR_OPEN_RANDOM);
```

--openssl ENGINE_add() code
```
/* Add another "ENGINE" type into the list. */
int ENGINE_add(ENGINE *e)
{
    int to_return = 1;
    if (e == NULL) {
        ERR_raise(ERR_LIB_ENGINE, ERR_R_PASSED_NULL_PARAMETER);
        return 0;
    }
    if ((e->id == NULL) || (e->name == NULL)) {
        ERR_raise(ERR_LIB_ENGINE, ENGINE_R_ID_OR_NAME_MISSING);
        return 0;
    }
    if (!CRYPTO_THREAD_write_lock(global_engine_lock))
        return 0;
    if (!engine_list_add(e)) {
        ERR_raise(ERR_LIB_ENGINE, ENGINE_R_INTERNAL_LIST_ERROR);
        to_return = 0;
    }
    CRYPTO_THREAD_unlock(global_engine_lock);
    return to_return;
}

```
Based on the s2n-tls code the e is not NULL. Its name and id are also set. And e is added for the first time. So the possible line that introduces the error should be CRYPTO_THREAD_write_lock(global_engine_lock). Checked openssl code the global_engine_lock seems an empty pointer here. So I want to know what could lead global_engine_lock to be a NULL. It works for more than one month. Now we have not changed anything, why does it suddenly not work? Is it a github host related issue or openssl bug? Please give me a help as it is a blocking issue for us. Thanks in advance!

My openssl version is 3.0.5.
OS: Ubuntu 20.04.

**Comment by paulidale** [comment]
Have you confirmed that `global_engine_lock` is NULL or do you just suspect it might be?

`global_engine_lock` should be initialised by the `ENGINE_new()` call if it hasn't already been set.  This call fails if it cannot initialise `global_engine_lock`.

Have you tried printing the OpenSSL errors out?  `ERR_print_errors_fp(stderr)` does this.

**Comment by RoyZhang2022** [comment]
Thanks for your reply. I just suspect as I do not have openssl detail info. I just analyze based on the caller s2n code. Is there a way to get the more error details from caller?

**Comment by paulidale** [comment]
The `ERR_print_errors_fp(stderr)` call will print the OpenSSL error stack to stderr.  If that's not possible, look at `ERR_print_errors_cb` to set up a callback to some code that can capture & log the error stack.

**Comment by RoyZhang2022** [comment]
Thanks Pauli! Let me add this to get more details.

**Comment by RoyZhang2022** [comment]
I got this error

error:13000067:engine routines::conflicting engine id

**Comment by paulidale** [comment]
That error is only raised in two places in the code.  I suspect it's the one in `engine_list_add()` that you're hitting.
This is triggered if two engines have the same ID.

Is this engine being loaded elsewhere?  Via configuration perhaps?

**Comment by RoyZhang2022** [comment]
From the print code I added to s2n I see it is using openssl3.0.5. Is it possible to have a bug here?

**Comment by RoyZhang2022** [comment]
By adding debug info to function that calls ENGINE_remove(), I see there is one ENGINE_remove is not executed. Need to investigate the root cause. It is related with our multi-threads execution but we have already protection for only once Aws::InitAPI() and Aws::ShutdownAPI().

**Comment by paulidale** [comment]
I wouldn't be at all surprised if engine load/unload isn't thread safe.

**Comment by nhorman** [comment]
marking as inactive, will be closed at the end of 3.4 dev barring further input
